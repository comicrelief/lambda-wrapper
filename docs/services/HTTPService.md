# HTTPService

Wrapper for `axios.request` that:

- sets a default timeout of 10 seconds
- forwards a `x-comicrelief-test-metadata` header if one was provided in the request from upstream
