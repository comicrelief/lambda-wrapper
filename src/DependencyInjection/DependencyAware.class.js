/**
 * DependencyAwareClass Class
 */
export default class DependencyAwareClass {
  /**
   * DependencyAwareClass constructor
   * @param di DependencyInjection
   */
  constructor(di) {
    this.di = di;
  }

  /**
   * Get Dependency Injection Container
   * @return DependencyInjection
   */
  getContainer() {
    return this.di;
  }
}
