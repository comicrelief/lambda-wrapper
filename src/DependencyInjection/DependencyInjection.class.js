import { DEFINITIONS, DEPENDENCIES } from '../Config/Dependencies';

/**
 * DependencyInjection class
 */
export default class DependencyInjection {
  /**
   * DependencyInjection constructor
   * @param configuration
   * @param event
   * @param context
   */
  constructor(configuration, event, context) {
    this.event = event;
    this.context = context;

    this.dependencies = {};
    this.configuration = configuration;

    for (let x = 0; x <= 1; x += 1) {
      // Iterate over lapper dependencies and add to container
      Object.keys(DEFINITIONS).forEach((dependencyKey) => {
        this.dependencies[dependencyKey] = new DEPENDENCIES[dependencyKey](this);
      });

      // Iterate over child dependencies and add to container
      if (typeof configuration.DEPENDENCIES !== 'undefined') {
        Object.keys(configuration.DEPENDENCIES).forEach((dependencyKey) => {
          this.dependencies[dependencyKey] = new configuration.DEPENDENCIES[dependencyKey](this);
        });
      }
    }
  }

  /**
   * Get Dependency
   * @param definition
   * @return {*}
   */
  get(definition) {
    if (typeof this.dependencies[definition] === 'undefined') {
      throw Error(`${definition} does not exist in di container`);
    }

    return this.dependencies[definition];
  }

  /**
   * Get Event
   * @return {*}
   */
  getEvent() {
    return this.event;
  }

  /**
   * Get Context
   * @return {*}
   */
  getContext() {
    return this.context;
  }

  /**
   * Get Configuration
   * @param definition string
   * @return {*}
   */
  getConfiguration(definition = null) {
    if (definition !== null && typeof this.configuration[definition] === 'undefined') {
      return null;
    } else if (typeof this.configuration[definition] !== 'undefined') {
      return this.configuration[definition];
    }

    return this.configuration;
  }
}
