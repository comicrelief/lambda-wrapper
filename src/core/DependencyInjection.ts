import { Context } from 'aws-lambda';

import DependencyAwareClass from './DependencyAwareClass';
import { LambdaWrapperConfig } from './config';

// eslint-disable-next-line no-use-before-define
type Class<TInstance, TConfig extends LambdaWrapperConfig> = new (di: DependencyInjection<TConfig>) => TInstance;

/**
 * Dependency injection container.
 *
 * Dependencies (singleton instances of dependency-aware classes) are provided
 * to the main Lambda handler and other dependencies via this class.
 */
export default class DependencyInjection<TConfig extends LambdaWrapperConfig = any> {
  /**
   * Instantiated dependencies.
   */
  readonly dependencies: Record<string, DependencyAwareClass>;

  /**
   * True until all dependencies have been constructed.
   */
  private isConstructing = true;

  constructor(
    readonly config: TConfig,
    readonly event: any,
    readonly context: Context,
  ) {
    // get unique dependency classes -- a class may be included several times,
    // but should be instantiated only once
    const classes = Array.from(new Set(Object.values(config.dependencies)));

    // guard against duplicate keys
    const countByName = classes
      .map((Constructor) => Constructor.name)
      .reduce(
        (counts, name) => ({ ...counts, [name]: (counts[name] || 0) + 1 }),
        {} as Record<string, number>,
      );
    if (Object.values(countByName).some((count) => count > 1)) {
      const duplicateNames = Object.entries(countByName)
        .filter(([, count]) => count > 1)
        .map(([name]) => name);

      // if all class names are single-letter, they're probably minified -- in
      // this case, give a hint about how to fix it
      const action = duplicateNames.every((it) => it.length === 1)
        ? "If you don't recognise the single-letter names listed above, your "
          + "bundler may be minifying your code. You'll need to disable this "
          + 'for Lambda Wrapper to work correctly. Please refer to the Notes '
          + 'section of the Lambda Wrapper readme:\n\n'
          + '  https://github.com/comicrelief/lambda-wrapper#notes'
        : 'Please ensure that all dependency classes have a unique name.';

      throw new Error(
        `Dependency names are not unique: ${duplicateNames.join(', ')}\n\n${action}`,
      );
    }

    // instantiate all dependencies
    this.dependencies = Object.fromEntries(
      classes.map((Constructor) => [Constructor.name, new Constructor(this)]),
    );

    this.isConstructing = false;
  }

  /**
   * Get the singleton instance of the given dependency.
   *
   * @param dependency
   */
  get<T extends DependencyAwareClass>(dependency: Class<T, TConfig>): T {
    if (this.isConstructing) {
      throw new Error(
        'Dependencies are not available in dependency class constructors.\n\n'
        + 'To fix this, call `di.get` in the function where the dependency is '
        + 'used instead of inside your constructor.',
      );
    }

    const name = dependency.name;

    if (!this.dependencies[name]) {
      throw new Error(
        `${name} does not exist in dependency container\n\n`
        + `Make sure you've included ${name} in the 'dependencies' key of your `
        + 'Lambda Wrapper config.',
      );
    }

    return this.dependencies[name] as T;
  }

  /**
   * Get the event passed to AWS Lambda.
   *
   * @deprecated Use `di.event` instead.
   */
  getEvent() {
    return this.event;
  }

  /**
   * Get the AWS Lambda context object.
   *
   * @deprecated Use `di.context` instead.
   */
  getContext() {
    return this.context;
  }

  /**
   * Get Lambda Wrapper configuration.
   *
   * @deprecated Use `di.config` instead.
   */
  getConfiguration() {
    return this.config;
  }

  /**
   * True if the function is being executed in `serverless-offline`.
   *
   * We use the following checks for this:
   *
   * - if there is no function ARN, or the ARN includes 'offline'
   * - if `process.env.USE_SERVERLESS_OFFLINE` is set
   *
   * TODO: This is nothing to do with dependency injection and should be moved
   *   somewhere else! Any ideas?
   */
  get isOffline(): boolean {
    return !this.context.invokedFunctionArn
      || this.context.invokedFunctionArn.includes('offline')
      || !!process.env.USE_SERVERLESS_OFFLINE;
  }

  /**
   * Get the `service-stage-` prefix added by Serverless to deployed Lambda
   * function names. This is handy when you want to invoke other functions,
   * without having to hardcode the service name and stage.
   *
   * The returned prefix includes a trailing dash. To get the deployed name of
   * another Lambda function, concatenate its serverless function name (its key
   * in `serverless.yml`) onto the prefix:
   *
   * ```js
   * const serverlessFunctionName = 'MyFunction';
   * const deployedName = `${di.getLambdaPrefix()}${serverlessFunctionName}`;
   * ```
   *
   * This function relies on looking at the currently running Lambda function's
   * resource name. It will not work correctly if the Lambda function has been
   * given a custom resource name.
   */
  getLambdaPrefix(): string {
    const stage = process.env.STAGE;
    if (!stage) {
      /* eslint-disable no-template-curly-in-string */
      throw new Error(
        'STAGE is not set\n\n'
        + 'Please add to your Lambda environment:\n\n'
        + '    STAGE: ${sls:stage}\n',
      );
    }
    if (!this.context.functionName) {
      throw new Error('Lambda function name is unavailable in context');
    }
    const [service] = this.context.functionName.split(`-${stage}-`);
    return `${service}-${stage}-`;
  }
}
