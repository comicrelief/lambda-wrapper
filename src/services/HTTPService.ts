import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';

import DependencyAwareClass from '../core/DependencyAwareClass';
import DependencyInjection from '../core/DependencyInjection';
import RequestService from './RequestService';

export const COMICRELIEF_TEST_METADATA_HEADER = 'x-comicrelief-test-metadata';
export const DEFAULT_HTTP_TIMEOUT = 10 * 1000;

/**
 * Wrapper for `axios.request` that:
 *
 * - sets a default timeout of 10 seconds
 * - forwards a `x-comicrelief-test-metadata` header if one was provided in
 *   the request from upstream
 */
export default class HTTPService extends DependencyAwareClass {
  config: AxiosRequestConfig;

  constructor(di: DependencyInjection) {
    super(di);

    this.config = {
      timeout: DEFAULT_HTTP_TIMEOUT,
    };
  }

  /**
   * Set the default timeout.
   *
   * @param ms
   */
  setDefaultTimeout(ms: number) {
    this.config.timeout = ms;
  }

  /**
   * Perform an HTTP request.
   *
   * @param config
   */
  async request(config: AxiosRequestConfig): Promise<AxiosResponse> {
    const mergedConfig = {
      timeout: this.config.timeout,
      headers: {},
      ...config,
    };

    const request = this.di.get(RequestService);
    const testMetadata = request.getHeader(COMICRELIEF_TEST_METADATA_HEADER);

    if (testMetadata) {
      mergedConfig.headers = mergedConfig.headers || {};
      mergedConfig.headers[COMICRELIEF_TEST_METADATA_HEADER] = testMetadata;
    }

    return axios.request(mergedConfig);
  }
}
