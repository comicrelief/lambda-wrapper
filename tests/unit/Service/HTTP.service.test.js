import axios from 'axios';

import CONFIGURATION from '../../../src/Config/Dependencies';
import DependencyInjection from '../../../src/DependencyInjection/DependencyInjection.class';
import HTTPService, { COMICRELIEF_TEST_METADATA_HEADER } from '../../../src/Service/HTTP.service';
import getEvent from '../../mocks/aws/event.json';

const getContext = { invokedFunctionArn: 'my-function' };

const getService = (event = getEvent, context = getContext) => new HTTPService(new DependencyInjection(CONFIGURATION, event, context));

describe('Service/HTTPService', () => {
  afterEach(() => jest.clearAllMocks());

  describe('request', () => {
    const testCases = {
      'GET Request': { method: 'GET', url: '/' },
      'POST Request': { method: 'POST', url: '/' },
      'PUT Request': { method: 'PUT', url: '/' },
      'PATCH Request': { method: 'PATCH', url: '/' },
      'HEAD Request': { method: 'HEAD', url: '/' },
      'DELETE Request': { method: 'DELETE', url: '/' },
      'with URL': { url: '/some/nested/path' },
      'with baseURL': { baseUrl: 'https://comicrelief.com/test', url: '/additional/url' },
      'overriding timeout': { timeout: 99 },
      'with headers': { headers: { Authorization: 'Bearer test' } },
      'with undefined headers': { headers: undefined },
    };

    Object.entries(testCases).forEach(([description, config]) => {
      it(description, async () => {
        const expected = { response: {} };
        const mock = jest.spyOn(axios, 'request').mockResolvedValue(expected);
        const service = getService();

        const response = await service.request(config);

        expect(response).toEqual(expected);
        expect(mock.mock.calls).toMatchSnapshot('config');
      });

      it(`adds the test header, ${description}`, async () => {
        const metadata = JSON.stringify({ user: 'Dante Alighieri' });
        const event = {
          ...getEvent,
          headers: {
            ...getEvent.headers,
            [COMICRELIEF_TEST_METADATA_HEADER]: metadata,
          },
        };
        const expected = { response: {} };
        const mock = jest.spyOn(axios, 'request').mockResolvedValue(expected);
        const service = getService(event);

        const response = await service.request(config);

        expect(response).toEqual(expected);
        expect(mock.mock.calls).toMatchSnapshot('config');
      });
    });
  });
});
