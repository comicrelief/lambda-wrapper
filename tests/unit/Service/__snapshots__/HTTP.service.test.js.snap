// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service/HTTPService request DELETE Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {},
      "method": "DELETE",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request GET Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {},
      "method": "GET",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request HEAD Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {},
      "method": "HEAD",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request PATCH Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {},
      "method": "PATCH",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request POST Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {},
      "method": "POST",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request PUT Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {},
      "method": "PUT",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request adds the test header, DELETE Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "x-comicrelief-test-metadata": "{\\"user\\":\\"Dante Alighieri\\"}",
      },
      "method": "DELETE",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request adds the test header, GET Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "x-comicrelief-test-metadata": "{\\"user\\":\\"Dante Alighieri\\"}",
      },
      "method": "GET",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request adds the test header, HEAD Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "x-comicrelief-test-metadata": "{\\"user\\":\\"Dante Alighieri\\"}",
      },
      "method": "HEAD",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request adds the test header, PATCH Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "x-comicrelief-test-metadata": "{\\"user\\":\\"Dante Alighieri\\"}",
      },
      "method": "PATCH",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request adds the test header, POST Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "x-comicrelief-test-metadata": "{\\"user\\":\\"Dante Alighieri\\"}",
      },
      "method": "POST",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request adds the test header, PUT Request: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "x-comicrelief-test-metadata": "{\\"user\\":\\"Dante Alighieri\\"}",
      },
      "method": "PUT",
      "timeout": 10000,
      "url": "/",
    },
  ],
]
`;

exports[`Service/HTTPService request adds the test header, overriding timeout: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "x-comicrelief-test-metadata": "{\\"user\\":\\"Dante Alighieri\\"}",
      },
      "timeout": 99,
    },
  ],
]
`;

exports[`Service/HTTPService request adds the test header, with URL: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "x-comicrelief-test-metadata": "{\\"user\\":\\"Dante Alighieri\\"}",
      },
      "timeout": 10000,
      "url": "/some/nested/path",
    },
  ],
]
`;

exports[`Service/HTTPService request adds the test header, with baseURL: config 1`] = `
Array [
  Array [
    Object {
      "baseUrl": "https://comicrelief.com/test",
      "headers": Object {
        "x-comicrelief-test-metadata": "{\\"user\\":\\"Dante Alighieri\\"}",
      },
      "timeout": 10000,
      "url": "/additional/url",
    },
  ],
]
`;

exports[`Service/HTTPService request adds the test header, with headers: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Authorization": "Bearer test",
        "x-comicrelief-test-metadata": "{\\"user\\":\\"Dante Alighieri\\"}",
      },
      "timeout": 10000,
    },
  ],
]
`;

exports[`Service/HTTPService request overriding timeout: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {},
      "timeout": 99,
    },
  ],
]
`;

exports[`Service/HTTPService request with URL: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {},
      "timeout": 10000,
      "url": "/some/nested/path",
    },
  ],
]
`;

exports[`Service/HTTPService request with baseURL: config 1`] = `
Array [
  Array [
    Object {
      "baseUrl": "https://comicrelief.com/test",
      "headers": Object {},
      "timeout": 10000,
      "url": "/additional/url",
    },
  ],
]
`;

exports[`Service/HTTPService request with headers: config 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Authorization": "Bearer test",
      },
      "timeout": 10000,
    },
  ],
]
`;
